File: compositor/src/main/java/org/westford/launch/indirect/IndirectLifeCycleSignalsFactory.java
package org.westford.launch.indirect;


import org.freedesktop.wayland.server.Display;

import javax.annotation.Nonnull;
import javax.inject.Inject;

import static org.freedesktop.wayland.server.jaccall.WaylandServerCore.WL_EVENT_READABLE;

public class IndirectLifeCycleSignalsFactory {

    @Nonnull
    private final PrivateIndirectLifeCycleSignalsFactory privateIndirectLifeCycleSignalsFactory;
    @Nonnull
    private final Display                                display;

    @Inject
    IndirectLifeCycleSignalsFactory(@Nonnull final PrivateIndirectLifeCycleSignalsFactory privateIndirectLifeCycleSignalsFactory,
                                    @Nonnull final Display display) {
        this.privateIndirectLifeCycleSignalsFactory = privateIndirectLifeCycleSignalsFactory;
        this.display = display;
    }

    public IndirectLifeCycleSignals create() {
        final int socketFd1 = Integer.parseInt(System.getenv(NativeConstants.ENV_WESTFORD_LAUNCHER_SOCK));

        final IndirectLifeCycleSignals indirectLifeCycleSignals = this.privateIndirectLifeCycleSignalsFactory.create(socketFd1);
        this.display.getEventLoop()
                    .addFileDescriptor(socketFd1,
                                       WL_EVENT_READABLE,
                                       indirectLifeCycleSignals::handleLauncherEvent);

        return indirectLifeCycleSignals;
    }
}

File: compositor/src/main/java/org/westford/launch/indirect/IndirectLifeCycleSignalsFactory.kt
package org.westford.launch.indirect


import org.freedesktop.wayland.server.Display
import javax.inject.Inject

import org.freedesktop.wayland.server.jaccall.WaylandServerCore.WL_EVENT_READABLE

class IndirectLifeCycleSignalsFactory @Inject
internal constructor(private val privateIndirectLifeCycleSignalsFactory: PrivateIndirectLifeCycleSignalsFactory,
                     private val display: Display) {

    fun create(): IndirectLifeCycleSignals {
        val socketFd1 = Integer.parseInt(System.getenv(NativeConstants.ENV_WESTFORD_LAUNCHER_SOCK))

        val indirectLifeCycleSignals = this.privateIndirectLifeCycleSignalsFactory.create(socketFd1)
        this.display.eventLoop
                .addFileDescriptor(socketFd1,
                        WL_EVENT_READABLE,
                        FileDescriptorEventHandler { fd, mask -> indirectLifeCycleSignals.handleLauncherEvent(fd, mask) })

        return indirectLifeCycleSignals
    }
}
